core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java

```diff
@@ -0,0 +1,251 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.pool.KeyedChannelPoolHandler;
+import com.linecorp.armeria.client.pool.PoolKey;
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.internal.TransportType;
+
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollChannelOption;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.dns.DnsAddressResolverGroup;
+import io.netty.resolver.dns.DnsServerAddressStreamProviders;
+
+@SuppressWarnings("checkstyle:JavadocMethod")
+public final class ClientFactoryBuilder {
+
+    private static final KeyedChannelPoolHandler<PoolKey> DEFAULT_CONNECTION_POOL_LISTENER =
+            KeyedChannelPoolHandler.noop();
+
+    private static final Consumer<SslContextBuilder> DEFAULT_SSL_CONTEXT_CUSTOMIZER = b -> { /* no-op */ };
+
+    private static final Function<EventLoopGroup, DnsAddressResolverGroup>
+            DEFAULT_ADDRESS_RESOLVER_GROUP_FACTORY = eventLoopGroup -> new DnsAddressResolverGroup(
+            TransportType.datagramChannelType(eventLoopGroup),
+            DnsServerAddressStreamProviders.platformDefault());
+
+    // Do not accept 1) the options that may break Armeria and 2) the deprecated options.
+    @SuppressWarnings("deprecation")
+    private static final Set<ChannelOption<?>> PROHIBITED_SOCKET_OPTIONS = ImmutableSet.of(
+            ChannelOption.ALLOW_HALF_CLOSURE, ChannelOption.AUTO_READ,
+            ChannelOption.AUTO_CLOSE, ChannelOption.MAX_MESSAGES_PER_READ,
+            ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK, ChannelOption.WRITE_BUFFER_LOW_WATER_MARK,
+            EpollChannelOption.EPOLL_MODE);
+
+    private static final String MESSAGE_MUTUALLY_EXCLUSIVE =
+            "eventLoopGroup() and useDaemonThreads/numWorkerThreads() are mutually exclusive.";
+
+    // Netty-related properties:
+    private Boolean createEventLoopGroup;
+    @Nullable
+    private EventLoopGroup eventLoopGroup; // Created and managed by ClientFactory if null.
```

Comment:
It turns out to be more complicated than expected because of how `Server` manages the connection life cycle. Let me send a PR for common event loop groups.